### PR TITLE
Coupons: Enable Save button on Edit Coupon screen only when there are changes made

### LINKED ISSUE #6832: Coupons: Only enable saving coupon details if any changes were made
Currently, we are showing the Save button on the Edit Coupon screen despite there having been no change made to the details.

We should hide the Save button in this case.

### PR DESCRIPTION
**Pull request by itsmeichigo**
Closes: #6832  

### Description

This PR adds logic to check for changes on the Edit Coupon screen, so that the Save button is only enabled if there are any changes made to the coupon details.

This also updates the way `populatedCoupon` is created - previously we're using `copy` to update values for the existing coupon. This method ignores nil values, so we cannot remove the expiry date for the coupon by sending nil to the method. The solution is to create a new Coupon manually instead.

### Testing instructions

- Make sure that you have Coupon Management enabled in Settings > Experimental Features.
- Navigate to Menu > Coupons > select a coupon > select the ellipsis button and choose Edit Coupon.
- Notice that the Save button is disabled initially. Editing any of the fields should enable the button.

### Screenshots

https://user-images.githubusercontent.com/5533851/168049177-7f7f1802-f76d-415f-bcf4-295159c88d09.mp4

---
- [x] I have considered if this change warrants user-facing release notes and have added them to `RELEASE-NOTES.txt` if necessary.

### CODE REVIEW

**Review by pmusolino — COMMENTED** [review]
I just noticed that with some coupons the button is always enabled, and for other coupons no.
I believe that we are changing some information under the hood, and we are not taking in consideration that changes 🤔 WDYT?

https://user-images.githubusercontent.com/495617/168105987-742ecf1d-92e7-41c9-af16-83e38f18e269.mp4

**Comment by itsmeichigo** [comment]
Hi @pmusolino - the issue seems to be caused by the trimming of time on the date when creating the `populatedCoupon`. I think it makes more sense to only do this when picking the expiry date, so I moved this logic to the date picker in dc374d2071a8d9bab37dc26ae3fdc749985c6eb8. Let me know what you think about this - thanks!

**Comment by pmusolino** [comment]
> Hi @pmusolino - the issue seems to be caused by the trimming of time on the date when creating the populatedCoupon. I think it makes more sense to only do this when picking the expiry date, so I moved this logic to the date picker in https://github.com/woocommerce/woocommerce-ios/commit/dc374d2071a8d9bab37dc26ae3fdc749985c6eb8. Let me know what you think about this - thanks!

It looks like that with this change, also If I change the date, it will be considered as the same day 💭 looks at the video below for an example of what's happening.

https://user-images.githubusercontent.com/495617/168123329-5966d8da-d636-48af-9cde-2b7a329e90ac.mp4

**Comment by itsmeichigo** [comment]
Good point! It's not immediately obvious to me why we have a different time for the expiry date - whether it was due to the previous expiry date being set when the store was configured with a different timezone?

However, I suspect since we're trimming time on the updated expiry date, we should trim the time of the initial one as well before comparing. I reverted the previous change and added the trimming in dcd137ef1ea1da21b0c1215e49c2d180ccb6d2af. This works well for different stores, so I hope this is a reliable fix.

Let me know what you think about the solution @pmusolino.

File: WooCommerce/Classes/ViewRelated/Coupons/Add and Edit Coupons/AddEditCouponViewModel.swift [hunk 1]
```diff
@@ -123,6 +123,18 @@ final class AddEditCouponViewModel: ObservableObject {
...
 
+    var hasChangesMade: Bool {
+        let coupon = populatedCoupon
+        return checkAmountUpdated(for: coupon) ||
+        checkDescriptionUpdated(for: coupon) ||
+        checkCouponCodeUpdated(for: coupon) ||
+        checkAllowedProductsAndCategoriesUpdated(for: coupon) ||
+        checkExpiryDateUpdated(for: coupon) ||
+        checkFreeShippingUpdated(for: coupon) ||
+        checkAllowedProductsAndCategoriesUpdated(for: coupon) ||
```
Comment (selanthiraiyan) [inline WooCommerce/Classes/ViewRelated/Coupons/Add and Edit Coupons/AddEditCouponViewModel.swift]: I think we are checking for `checkAllowedProductsAndCategoriesUpdated` twice.

File: WooCommerce/Classes/ViewRelated/Coupons/Add and Edit Coupons/AddEditCouponViewModel.swift [hunk 2]
```diff
@@ -335,23 +328,63 @@ private extension AddEditCouponViewModel {
...
-                return initialCoupon.dateExpires != coupon.dateExpires
-            }
-            return !oldDate.isSameDay(as: newDate)
-        }()
+    func checkExpiryDateUpdated(for coupon: Coupon) -> Bool {
+        guard let initialCoupon = self.coupon else {
+            return false
+        }
+        guard let oldDate = initialCoupon.dateExpires?.startOfDay(timezone: timezone),
+                let newDate = coupon.dateExpires else {
```
Comment (selanthiraiyan) [inline WooCommerce/Classes/ViewRelated/Coupons/Add and Edit Coupons/AddEditCouponViewModel.swift]: Nit: Indentation fix needed.

File: WooCommerce/Classes/ViewRelated/Coupons/Add and Edit Coupons/AddEditCouponViewModel.swift [hunk 3]
```diff
@@ -335,23 +328,63 @@ private extension AddEditCouponViewModel {
...
-            guard let oldDate = initialCoupon.dateExpires, let newDate = coupon.dateExpires else {
-                return initialCoupon.dateExpires != coupon.dateExpires
-            }
-            return !oldDate.isSameDay(as: newDate)
-        }()
+    func checkExpiryDateUpdated(for coupon: Coupon) -> Bool {
+        guard let initialCoupon = self.coupon else {
+            return false
+        }
+        guard let oldDate = initialCoupon.dateExpires?.startOfDay(timezone: timezone),
```
Comment (selanthiraiyan) [inline WooCommerce/Classes/ViewRelated/Coupons/Add and Edit Coupons/AddEditCouponViewModel.swift:338]: Would it make sense to add a comment to explain why we are trimming time here?

File: WooCommerce/WooCommerceTests/ViewRelated/Coupons/AddEditCouponViewModelTests.swift [hunk 4]
```diff
@@ -126,7 +126,126 @@ final class AddEditCouponViewModelTests: XCTestCase {
         XCTAssertNil(result)
     }
 
-    private enum Localization {
+    func test_hasChangesMade_is_correct_for_coupon_code() {
```
Comment (selanthiraiyan) [inline WooCommerce/WooCommerceTests/ViewRelated/Coupons/AddEditCouponViewModelTests.swift:129]: Nice tests. ❤️

**Review by selanthiraiyan — APPROVED** [review]
Nice work @itsmeichigo 🎉 

It tests well. ✅ 

I left a few non-blocker comments.

File: WooCommerce/Classes/ViewRelated/Coupons/Add and Edit Coupons/AddEditCouponViewModel.swift (hunk 3 above)
Comment (itsmeichigo) [reply WooCommerce/Classes/ViewRelated/Coupons/Add and Edit Coupons/AddEditCouponViewModel.swift:338]: Done here: a35e4646f8f2846dcf6805ee2fc18b151c316b3d

File: WooCommerce/Classes/ViewRelated/Coupons/Add and Edit Coupons/AddEditCouponViewModel.swift (hunk 2 above)
Comment (itsmeichigo) [reply WooCommerce/Classes/ViewRelated/Coupons/Add and Edit Coupons/AddEditCouponViewModel.swift]: Done here: bd852aeb47afc497e14442871341078561ee5698

File: WooCommerce/Classes/ViewRelated/Coupons/Add and Edit Coupons/AddEditCouponViewModel.swift (hunk 1 above)
Comment (itsmeichigo) [reply WooCommerce/Classes/ViewRelated/Coupons/Add and Edit Coupons/AddEditCouponViewModel.swift]: Great catch! Removed in c003d1f9a0e690015b4ab71c2368d02dec324671.

**Pull request merged**